### DISCUSSION TITLE
Version 4 plugins not working?

### DISCUSSION THREAD

**Discussion by cyberalien** [tailwindlabs/tailwindcss#15458] [Help]
I'm trying to port a plugin to version 4 to use `@plugin` directive, this is code I'm trying to make work:
```ts
/* eslint-disable @typescript-eslint/unbound-method */
import plugin from 'tailwindcss/plugin';

function testPlugin(): ReturnType<typeof plugin> {
	console.log('Default export is working');
	return plugin(({ matchComponents }) => {
		console.log('Plugin is initialised');
		matchComponents({
			foo: (value: string) => ({
				'--foo': value,
			}),
		});
	});
}

// Default export
export default testPlugin;
```

It is in monorepo, so one package for plugin, one package for demo. Using `@plugin 'test-plugin';` seems to work, but plugin does not work.

In logs I see:
```
Default export is working
```
So Tailwind does import plugin and it should work.

What I don't see in log is `Plugin is initialised`, so plugin is not actually being called.

Since I couldn't find any documentation on building plugins for version 4, with trial and error it looks like `import plugin from 'tailwindcss/plugin';` is sort of working (see below).

## Issues with code

There are few issues with code sample above though, so I'm not even sure if it is correct way to build a plugin:

1. TypeScript complains about return type of `testPlugin()`:

```
The inferred type of 'testPlugin' cannot be named without a reference to '../node_modules/tailwindcss/dist/types-BTRmm49E.mjs'. This is likely not portable. A type annotation is necessary.ts(2742)
```

2. ESLint complains about using `matchComponents`, so had to disable `@typescript-eslint/unbound-method` rule:

```
Avoid referencing unbound methods which may cause unintentional scoping of `this`.
If your function does not access `this`, you can annotate it with `this: void`, or consider using an arrow function instead.eslint[@typescript-eslint/unbound-method](https://typescript-eslint.io/rules/unbound-method)
```

3. Couldn't even build that file without setting `"moduleResolution": "nodenext",` in tsconfig, otherwise import for `tailwindcss/plugin` cannot be resolved.

All of that is very strange, which leads me to believe that I'm building plugin incorrectly and it should be imported from a different package, not from `tailwindcss/plugin`.

What am I doing wrong?

**Comment by silverwind** [comment]
https://github.com/tailwindlabs/tailwindcss/pull/19707 will solve it.
